File: block-chain-updated/core/relay-chain/api/pravah_adapter.go
package api

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"sort"
	"strconv"
	"sync"
	"time"
)

var (
	pravahURL   = getEnvOrDefault("PRAVAH_URL", "http://localhost:7000/api/runtime")
	ssplSecret  = getEnvOrDefault("SSPL_SECRET_KEY", "default-secret-key-change-in-prod")
	appName     = "block-chain-updated"
	heartOnce   sync.Once
	httpClient  = &http.Client{Timeout: 4 * time.Second}
)

func getEnvOrDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func signPayload(traceID, canonical string) (string, string) {
	bodyHashBytes := sha256.Sum256([]byte(canonical))
	bodyHash := hex.EncodeToString(bodyHashBytes[:])

	timestamp := strconv.FormatInt(time.Now().Unix(), 10)
	sigData := fmt.Sprintf("%s:%s:%s", traceID, timestamp, bodyHash)

	h := hmac.New(sha256.New, []byte(ssplSecret))
	h.Write([]byte(sigData))
	signature := hex.EncodeToString(h.Sum(nil))

	return timestamp, signature
}

func EmitPravahSignal(state string, latencyMs float64, errorsLastMin int) {
	go func() {
		defer func() {
			if r := recover(); r != nil {
				// Prevent crashes from panic
			}
		}()

		traceID := fmt.Sprintf("blockchain-%d", time.Now().UnixNano())

		payload := map[string]interface{}{
			"app":             appName,
			"env":             getEnvOrDefault("ENVIRONMENT", "dev"),
			"state":           state,
			"latency_ms":      latencyMs,
			"errors_last_min": errorsLastMin,
			"workers":         1,
		}

		// Sort keys for canonical JSON serialization
		keys := make([]string, 0, len(payload))
		for k := range payload {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		canonicalBuf := bytes.NewBuffer(nil)
		canonicalBuf.WriteString("{")
		for i, k := range keys {
			valBytes, _ := json.Marshal(payload[k])
			canonicalBuf.WriteString(fmt.Sprintf("%q:%s", k, string(valBytes)))
			if i < len(keys)-1 {
				canonicalBuf.WriteString(",")
			}
		}
		canonicalBuf.WriteString("}")

		canonicalStr := canonicalBuf.String()
		timestamp, signature := signPayload(traceID, canonicalStr)

		bodyBytes, _ := json.Marshal(payload)
		req, err := http.NewRequest("POST", pravahURL, bytes.NewBuffer(bodyBytes))
		if err != nil {
			return
		}

		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("X-Trace-Id", traceID)
		req.Header.Set("X-Timestamp", timestamp)
		req.Header.Set("X-Trace-Signature", signature)

		resp, err := httpClient.Do(req)
		if err == nil && resp != nil {
			resp.Body.Close()
		}
	}()
}

func StartHeartbeat(intervalSeconds int) {
	heartOnce.Do(func() {
		go func() {
			ticker := time.NewTicker(time.Duration(intervalSeconds) * time.Second)
			defer ticker.Stop()

			// Initial signal
			EmitPravahSignal("running", 0.0, 0)

			for range ticker.C {
				EmitPravahSignal("running", 0.0, 0)
			}
		}()
	})
}

File: bhiv-core-bucket-karma-prana-orchrstration-insightcore-insightflow/insightcore-bridgev4x-main/observability/pravah_adapter.py
import os
import uuid
import time
import hmac
import hashlib
import json
import logging
import threading

logger = logging.getLogger(__name__)

PRAVAH_URL = os.getenv("PRAVAH_URL", "http://localhost:7000/api/runtime")
SSPL_SECRET = os.getenv("SSPL_SECRET_KEY", "default-secret-key-change-in-prod")
APP_NAME = "bhiv-insight-core"

def _sign_payload(trace_id: str, canonical: str) -> tuple[str, str]:
    body_hash = hashlib.sha256(canonical.encode()).hexdigest()
    timestamp = str(int(time.time()))
    sig_data = f"{trace_id}:{timestamp}:{body_hash}"
    signature = hmac.new(SSPL_SECRET.encode(), sig_data.encode(), hashlib.sha256).hexdigest()
    return timestamp, signature

def _send(payload: dict) -> None:
    try:
        import requests as _req
    except ImportError:
        return

    trace_id = f"bhiv-{uuid.uuid4().hex[:16]}"
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    timestamp, signature = _sign_payload(trace_id, canonical)

    headers = {
        "X-Trace-Id": trace_id,
        "X-Timestamp": timestamp,
        "X-Trace-Signature": signature,
        "Content-Type": "application/json",
    }

    try:
        _req.post(PRAVAH_URL, json=payload, headers=headers, timeout=3)
    except Exception:
        pass

def emit_pravah_signal(
    state: str = "running",
    latency_ms: float = 0.0,
    errors_last_min: int = 0,
    workers: int = 1,
    extra: dict | None = None,
) -> None:
    payload = {
        "app": APP_NAME,
        "env": os.getenv("ENVIRONMENT", "dev"),
        "state": state,
        "latency_ms": int(latency_ms),
        "errors_last_min": int(errors_last_min),
        "workers": int(workers),
    }
    if extra:
        payload.update(extra)

    threading.Thread(target=_send, args=(payload,), daemon=True).start()

_heartbeat_running = False

def start_heartbeat(interval_seconds: int = 60) -> None:
    global _heartbeat_running
    if _heartbeat_running:
        return
    _heartbeat_running = True

    def _loop():
        while True:
            try:
                emit_pravah_signal(state="running")
            except Exception:
                pass
            time.sleep(interval_seconds)

    threading.Thread(target=_loop, daemon=True).start()
